server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -67,14 +71,23 @@ protected boolean resolveIndex(GetRequest request) {
 
     @Override
     protected ShardIterator shards(ClusterState state, InternalRequest request) {
-        return clusterService.operationRouting()
+        ShardIterator shards = clusterService.operationRouting()
             .getShards(
                 clusterService.state(),
                 request.concreteIndex(),
                 request.request().id(),
                 request.request().routing(),
                 request.request().preference()
             );
+        // If it is stateless, only route isPromotableToPrimary shards
+        if (DiscoveryNode.isStateless(clusterService.getSettings())) {
+            return new PlainShardIterator(
+                shards.shardId(),
+                shards.getShardRoutings().stream().filter(ShardRouting::isPromotableToPrimary).collect(Collectors.toList())
+            );
```

Comment:
Should this be inside of `getShards` to avoid duplication?
Also is this a temporary solution?

Reply:
This is temporary and I added a comment to this point. I did not move it in `getShards` as I felt like that surface level change was larger. And is relied on by other actions (explain and term vectors). However, I created an issue for those two transport actions, which would allow a more extensive refactoring in `getShards` if it makes sense.